File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int lastDigit;

int main(void)
{
    // Prompt user for credit card number.
    long cardNo = get_long("What's your credit card number? ");
    printf("You typed: %li\n", cardNo);

    // Finding length of an card number.
    int length = log10(cardNo) + 1;
    printf("Card length is: %i\n", length);

    // Is card length Even or odd?
    int Even = ((length / 2) * 2 == length) ? 1 : 0;
    printf("Has card even nubmers? %i\n", Even);

    // Find first 2 digits.
    long first2 = cardNo;
    if (Even == 1)
    {
        while (first2 >= 100)
        {
            first2 = first2 / 100;
        }
    }
    else
    {
        while (first2 >= 100)
        {
            first2 = first2 / 10;
        }
    }
    printf("First two numbers are: %li\n", first2);

    // Find first digit.
    int first1 = first2 / 10;
    printf("First digit is: %i\n", first1);

    // Get even positions of numbers from the card number
    long cardNoEven = cardNo / 10;   // Offsetting number by one digit to obtain even numbers
    int sumEven;
    printf("Even\n");
    while (cardNoEven > 0)
    {
        if ((cardNoEven % 10) < 5)
        {
            sumEven = sumEven + (cardNoEven % 10) * 2;
            printf("%li", (cardNoEven % 10) * 2);
        }
        else
        {
            int x = ((cardNoEven % 10) * 2);
            printf("%i", x);
            while (x > 0)
            {
                sumEven = sumEven + (x % 10);
                x = x / 10;
            }
        }
        cardNoEven = cardNoEven / 100;
    }
    printf("\nsum: %i\n", sumEven);

    // Get odd possitions of numbers from the card number
    long cardNoOdd = cardNo;
    int sumOdd;
    printf("Odd\n");
    while (cardNoOdd > 0)
    {
        sumOdd = sumOdd + (cardNoOdd % 10);
        printf("%li", cardNoOdd % 10);
        cardNoOdd = cardNoOdd / 100;
    }
    printf("\nsum: %i\n", sumOdd);

    // Final calculataion and checking
    int finalSum;
    finalSum = sumOdd + sumEven;
    lastDigit = (finalSum % 10);
    printf("Final sum is: %i\n", finalSum);
    printf("Last digit is: %i\n", lastDigit);

    // Check if card is American express, Master Card of Visa.
    if (length == 15 && (first2 == 34 || first2 == 37) && lastDigit == 0)
    {
        printf("AMEX\n");
    }
    else if ((first2 >= 51 && first2 <= 55) && length == 16 && lastDigit == 0)
    {
        printf("MASTERCARD\n");
    }
    else if (first1 == 4 && (length == 13 || length == 16) && lastDigit == 0)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: problem-sets/pset6/sentimental-credit/credit.py
# TODO

from cs50 import get_int
# import re


def main():
    # Prompt user for credit card number.
    card_number = get_int("What's your credit card number? ")
    print("You typed: ", card_number)

    # Finding the length of an card number.
    card_length = len(str(card_number))
    print("Card length is: ", card_length)

    # Is card length Even or odd?
    is_even = True if card_length % 2 == 0 else False
    print("Is card length even?", is_even)

    # Get the first digit
    first_digit = int(str(card_number)[:1])
    print(first_digit)

    # Get the first two digits
    first_two_digits = int(str(card_number)[:2])
    print(first_two_digits)

    # Get even positions of numbers from the card number
    even_digits = str(card_number)[card_length - 2::-2]
    print("even: ", even_digits)

    even_digits_multiplied = ""
    for number in even_digits:
        number = int(number) * 2
        even_digits_multiplied += str(number)

    sum_of_even_digits = 0
    for number in even_digits_multiplied:
        sum_of_even_digits += int(number)
    print("even_digits_multiplied: ", even_digits_multiplied)
    print("sum_of_even_digits: ", sum_of_even_digits)

    # Get odd possitions of numbers from the card number
    odd_digits = str(card_number)[card_length - 1::-2]
    print("odd:", odd_digits)
    sum_of_odd_digits = 0
    for number in odd_digits:
        sum_of_odd_digits += int(number)

    # Final calculataion and check
    final_sum = sum_of_even_digits + sum_of_odd_digits
    last_digit = final_sum % 10
    print("Final sum of even digits:", sum_of_even_digits)
    print("Final sum of odd digits:", sum_of_odd_digits)
    print("Final sum is: ", final_sum)
    print("Final digit is:", last_digit)

    # Check if card is American express, Master Card of Visa.
    if (card_length == 15 and first_two_digits in (34, 37) and last_digit == 0):
        print("AMEX")
    elif (card_length == 16 and first_two_digits in range(51, 56) and last_digit == 0):
        print("MASTERCARD")
    elif (card_length in (13, 16) and first_digit == 4 and last_digit == 0):
        print("VISA")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()
